<div style="background-color:#0A1A2E; border-left:7px solid #2563EB; padding:25px; border-radius:14px;">
<h1 style="color:#FFFFFF; font-size:34px; text-shadow:0 0 6px #2563EB; margin:0; border-bottom: 3px solid #2563EB; padding-bottom: 8px;">
AI & ML Course with BinX — Week 1 — Day 4 — Pandas: Tabular Data
</h1>
</div>

<div style="background-color:#0A1A2E; border-left:5px solid #2563EB; padding:15px 20px; border-radius:10px;">
<h2 style="color:#FFFFFF; font-size:22px; text-shadow:0 0 5px #2563EB; margin:0;">
📊 About the Dataset
</h2>
<p style="color:#93C5FD; font-size:16px; line-height:1.6; margin-top:10px;">
This notebook uses a global CO2 emissions dataset containing yearly
emission values for countries and aggregated regions (e.g. "World",
"High income", "OECD members"). Each row represents one country/region
for a given year, with columns for the country code, country name,
year, and the recorded CO2 emission value.
</p>
</div>

<div style="background-color:#0A1A2E; border-left:5px solid #2563EB; padding:15px 20px; border-radius:10px;">
<h2 style="color:#FFFFFF; font-size:24px; text-shadow:0 0 5px #2563EB; margin:0;">
4.1 Series and DataFrames
</h2>
<p style="color:#93C5FD; font-size:16px; line-height:1.6; margin-top:10px;">
Quick example showing the difference between a Series and a DataFrame.
</p>
</div>

In [1]:
import pandas as pd

# Series: single labeled column
series = pd.Series([1, 2, 3], name="Example")
print("\033[94mSeries:\033[0m")
print(series)
print("\033[92mType:\033[0m", type(series))

# DataFrame: full table (columns and rows)
dataframe = pd.DataFrame({"even": [2, 4, 6], "odd": [1, 3, 5]})
print("\n\033[94mDataFrame:\033[0m")
print(dataframe)
print("\033[92mType:\033[0m", type(dataframe))

Series:
0    1
1    2
2    3
Name: Example, dtype: int64
Type: <class 'pandas.Series'>

DataFrame:
   even  odd
0     2    1
1     4    3
2     6    5
Type: <class 'pandas.DataFrame'>


<div style="background-color:#0A1A2E; border-left:5px solid #2563EB; padding:15px 20px; border-radius:10px;">
<h2 style="color:#FFFFFF; font-size:24px; text-shadow:0 0 5px #2563EB; margin:0;">
4.2 Loading and Inspecting Data
</h2>
<p style="color:#93C5FD; font-size:16px; line-height:1.6; margin-top:10px;">
Before doing anything with a dataset, the first step is always to load
it and take a first look — its shape, its columns, and a sample of the
actual rows.
</p>
</div>

In [6]:
import pandas as pd

df = pd.read_csv("co2_emissions_data.csv")

df.head().style.set_table_styles([
    {'selector': 'th', 'props': [('background-color', '#2563EB'), 
                                    ('color', 'white'), 
                                    ('padding', '8px'),
                                    ('border', '1px solid #93C5FD')]},
    {'selector': 'td', 'props': [('background-color', '#0A1A2E'), 
                                    ('color', 'white'), 
                                    ('padding', '8px'),
                                    ('border', '1px solid #93C5FD')]}
])

,country_code,country_name,year,value
0,ABW,Aruba,1960,11092.675000
1,ABW,Aruba,1961,11576.719000
2,ABW,Aruba,1962,12713.489000
3,ABW,Aruba,1963,12178.107000
4,ABW,Aruba,1964,11840.743000


In [8]:
print("\033[94mShape:\033[0m", df.shape)

Shape: (13953, 4)


In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 13953 entries, 0 to 13952
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   country_code  13953 non-null  str    
 1   country_name  13953 non-null  str    
 2   year          13953 non-null  int64  
 3   value         13953 non-null  float64
dtypes: float64(1), int64(1), str(2)
memory usage: 436.2 KB


In [6]:
df.describe()

,year,value
count,13953.000000,1.395300e+04
mean,1990.732316,8.254983e+05
std,17.187585,2.788923e+06
min,1960.000000,-8.067400e+01
25%,1976.000000,1.100000e+03
50%,1992.000000,1.390000e+04
75%,2006.000000,1.642779e+05
max,2019.000000,3.434401e+07


<div style="background-color:#0A1A2E; border-left:5px solid #2563EB; padding:15px 20px; border-radius:10px;">
<h2 style="color:#FFFFFF; font-size:24px; text-shadow:0 0 5px #2563EB; margin:0;">
4.3 Selecting and Filtering
</h2>
</div>

<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Select Single
 columns</h3>

In [7]:
df.columns

Index(['country_code', 'country_name', 'year', 'value'], dtype='str')

<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Select multiple columns</h3>

In [8]:
df[["country_code", "country_name", "value"]]

,country_code,country_name,value
0,ABW,Aruba,11092.675000
1,ABW,Aruba,11576.719000
2,ABW,Aruba,12713.489000
3,ABW,Aruba,12178.107000
4,ABW,Aruba,11840.743000
...,...,...,...
13948,ZWE,Zimbabwe,12430.000305
13949,ZWE,Zimbabwe,11020.000458
13950,ZWE,Zimbabwe,10340.000153
13951,ZWE,Zimbabwe,12380.000114


<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Check the range of value first to choose a logical threshold</h3>

In [9]:
df["value"].describe

<bound method NDFrame.describe of 0        11092.675000
1        11576.719000
2        12713.489000
3        12178.107000
4        11840.743000
             ...     
13948    12430.000305
13949    11020.000458
13950    10340.000153
13951    12380.000114
13952    11760.000229
Name: value, Length: 13953, dtype: float64>

<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Boolean filtering (rows where value is above the mean)</h3>

In [10]:
threshold = df["value"].mean()
df[df["value"] > threshold]

,country_code,country_name,year,value
423,ARB,Arab World,1996,8.436291e+05
424,ARB,Arab World,1997,8.847770e+05
425,ARB,Arab World,1998,8.897156e+05
426,ARB,Arab World,1999,9.016245e+05
427,ARB,Arab World,2000,9.391917e+05
...,...,...,...,...
13647,WLD,World,2015,3.299554e+07
13648,WLD,World,2016,3.301856e+07
13649,WLD,World,2017,3.351454e+07
13650,WLD,World,2018,3.428935e+07


<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Using .loc — country_name and year for rows above threshold</h3>

In [11]:
df.loc[df["value"]>threshold, ["country_name","year"]]

,country_name,year
423,Arab World,1996
424,Arab World,1997
425,Arab World,1998
426,Arab World,1999
427,Arab World,2000
...,...,...
13647,World,2015
13648,World,2016
13649,World,2017
13650,World,2018


<div style="background-color:#0A1A2E; border-left:5px solid #2563EB; padding:15px 20px; border-radius:10px;">
<h2 style="color:#FFFFFF; font-size:24px; text-shadow:0 0 5px #2563EB; margin:0;">
4.4 Cleaning Data
</h2>
<p style="color:#93C5FD; font-size:16px; line-height:1.6; margin-top:10px;">
Real data is rarely perfect — missing values, duplicates, and wrong
types are common. Cleaning it correctly is usually the biggest part of
any project, and everything downstream depends on doing it right.
</p>
</div>

count missing values per column

In [11]:
print("\033[94mMissing values per column:\033[0m")
df.isnull().sum()

Missing values per column:


country_code    0
country_name    0
year            0
value           0
dtype: int64

In [12]:
missing_count = df["value"].isnull().sum()
print("\033[93mMissing values in 'value' column:\033[0m", missing_count)

df["value"] = df["value"].fillna(df["value"].mean())
print("\033[92mFilled missing values with the mean.\033[0m")

Missing values in 'value' column: 0
Filled missing values with the mean.


In [13]:
print("Number of duplicate rows :", df.duplicated().sum())

Number of duplicate rows : 0


<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Remove Duplicates if any exist </h3>

In [14]:
df.drop_duplicates(inplace=True)
print("Dataset after remove duplicates - Shape-",df.shape)

Dataset after remove duplicates - Shape- (13953, 4)


<div style="background-color:#0A1A2E; border-left:5px solid #2563EB; padding:15px 20px; border-radius:10px;">
<h2 style="color:#FFFFFF; font-size:24px; text-shadow:0 0 5px #2563EB; margin:0;">
4.5 Grouping and Aggregation
</h2>
<p style="color:#93C5FD; font-size:16px; line-height:1.6; margin-top:10px;">
Average CO2 value per country.
</p>
</div>

In [13]:
print("\033[94mAverage CO2 value per country:\033[0m")
df.groupby("country_name")["value"].mean()

Average CO2 value per country:


country_name
Afghanistan                    2.505611e+03
Africa Eastern and Southern    3.463568e+05
Africa Western and Central     1.075567e+05
Albania                        4.413947e+03
Algeria                        7.031805e+04
                                   ...     
West Bank and Gaza             1.754320e+03
World                          2.195928e+07
Yemen, Rep.                    8.619050e+03
Zambia                         3.369701e+03
Zimbabwe                       1.116555e+04
Name: value, Length: 256, dtype: float64

<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Multiple aggregations per country (mean and max value)</h3>

In [14]:
print("\033[92mMean and max value per country:\033[0m")
df.groupby("country_name")["value"].agg(["mean", "max"])

Mean and max value per country:


,mean,max
country_name,,
Afghanistan,2.505611e+03,8.930000e+03
Africa Eastern and Southern,3.463568e+05,6.030300e+05
Africa Western and Central,1.075567e+05,2.207400e+05
Albania,4.413947e+03,8.984150e+03
Algeria,7.031805e+04,1.712500e+05
...,...,...
West Bank and Gaza,1.754320e+03,3.234294e+03
World,2.195928e+07,3.434401e+07
"Yemen, Rep.",8.619050e+03,2.783000e+04


<div style="background-color:#0A1A2E; border-left:7px solid #2563EB; padding:20px; border-radius:14px;">
<h1 style="color:#FFFFFF; font-size:30px; text-shadow:0 0 6px #2563EB; margin:0;">
🧪 Hands-On Lab: Exploring a Real Dataset
</h1>
</div>

<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Step 1: Load a CSV dataset and print shape, columns, dtypes</h3>

In [15]:
lab_df = pd.read_csv("co2_emissions_data.csv")

print("\033[94mShape:\033[0m", lab_df.shape)
print("\033[92mColumns:\033[0m", lab_df.columns.tolist())
print("\033[93mData types:\033[0m")
print(lab_df.dtypes)

Shape: (13953, 4)
Columns: ['country_code', 'country_name', 'year', 'value']
Data types:
country_code        str
country_name        str
year              int64
value           float64
dtype: object


<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Step 2: Count missing values and handle them (with justification)</h3>

In [16]:
print("\033[94mMissing values per column:\033[0m")
lab_df.isnull().sum()

Missing values per column:


country_code    0
country_name    0
year            0
value           0
dtype: int64

<div style="background-color:#0A1A2E; border-left:5px solid #2563EB; padding:15px 20px; border-radius:10px;">
<p style="color:#93C5FD; font-size:16px; line-height:1.6; margin:0;">
<b>Justification:</b> The dataset has zero missing values in all columns
(country_code, country_name, year, value). No filling or dropping is
needed — the data is already complete and ready for analysis.
</p>
</div>

<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Step 3: Filter the data to a meaningful subset and describe what you find</h3>

In [17]:
lab_threshold = lab_df["value"].mean()
high_emission = lab_df[lab_df["value"] > lab_threshold]

print("\033[92mThreshold (mean value):\033[0m", lab_threshold)
print("\033[94mRows above threshold:\033[0m", high_emission.shape[0], "out of", lab_df.shape[0])
high_emission.head()

Threshold (mean value): 825498.3433543197
Rows above threshold: 1727 out of 13953


,country_code,country_name,year,value
423,ARB,Arab World,1996,843629.050531
424,ARB,Arab World,1997,884777.026039
425,ARB,Arab World,1998,889715.587754
426,ARB,Arab World,1999,901624.461624
427,ARB,Arab World,2000,939191.728434


In [18]:
print("Total rows in lab_df:", lab_df.shape[0])
print("Rows above threshold:", high_emission.shape[0])

Total rows in lab_df: 13953
Rows above threshold: 1727


<div style="background-color:#0A1A2E; border-left:5px solid #2563EB; padding:15px 20px; border-radius:10px;">
<p style="color:#93C5FD; font-size:16px; line-height:1.6; margin:0;">
<b>Findings:</b> Filtering the dataset to records above the mean value
shows 1727 out of 13953 total rows represent higher-than-average CO2
emissions. This subset highlights which countries/years contribute most
to global emissions, with entities like "Arab World" appearing among the
higher-emission records.
</p>
</div>

<h3 style="color:#FFFFFF; text-shadow:0 0 5px #2563EB;">Step 4: Use groupby to compute an aggregate statistic per category and interpret it</h3>

In [19]:
country_avg = lab_df.groupby("country_name")["value"].mean().sort_values(ascending=False)
print("\033[92mTop 10 countries by average CO2 value:\033[0m")
country_avg.head(10)

Top 10 countries by average CO2 value:


country_name
World                        2.195928e+07
High income                  1.112143e+07
IDA & IBRD total             1.111222e+07
OECD members                 1.089645e+07
IBRD only                    1.066460e+07
Low & middle income          1.065039e+07
Post-demographic dividend    1.061569e+07
Middle income                1.049463e+07
Upper middle income          8.038619e+06
Late-demographic dividend    7.788907e+06
Name: value, dtype: float64

<div style="background-color:#0A1A2E; border-left:5px solid #2563EB; padding:15px 20px; border-radius:10px;">
<p style="color:#93C5FD; font-size:16px; line-height:1.6; margin:0;">
<b>Interpretation:</b> The top 10 entities above show the highest average
CO2 values in the dataset, dominated by large aggregated regions rather
than individual countries — highlighting the need to separate country
level records from group-level ones for precise analysis.
</p>
</div>